### ISSUE TITLE
ERROR: error reading file sanity.json

### ISSUE THREAD

**Issue by erikd** [bos/criterion#107]
After @ndmitchell 's patch to remove the embedded `js-jquery` and `js-flot` libraries, the tests fail for me (on one of two machines) with:

```
$ dist/build/sanity/sanity
benchmarking fib/fib 10
time                 2.858 μs   (2.854 μs .. 2.863 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.861 μs   (2.856 μs .. 2.867 μs)
std dev              17.83 ns   (13.84 ns .. 23.61 ns)

benchmarking fib/fib 22
time                 2.034 ms   (2.028 ms .. 2.040 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.050 ms   (2.045 ms .. 2.060 ms)
std dev              23.57 μs   (17.58 μs .. 35.98 μs)

benchmarking length . filter/string
time                 3.206 μs   (3.203 μs .. 3.210 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 3.207 μs   (3.204 μs .. 3.211 μs)
std dev              12.23 ns   (9.303 ns .. 16.52 ns)

benchmarking length . filter/bytestring
time                 892.5 ns   (890.2 ns .. 896.1 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 897.3 ns   (894.3 ns .. 902.2 ns)
std dev              13.40 ns   (7.963 ns .. 22.55 ns)
variance introduced by outliers: 15% (moderately inflated)

sanity: [Failed]
ERROR: error reading file sanity.json:
  "Error in $: endOfInput"
```

The two machines are pretty much identical, running Debian Testing on x86_64.

**Comment by ndmitchell** [comment]
When you say pretty much identical, are there any Haskell library version differences between the two?  That would be a good hint at what might have gone wrong.

**Comment by erikd** [comment]
I did a `cabal update` and rebuilt on the machine that was working and now it fails like the other machine.

**Comment by ndmitchell** [comment]
I can't reproduce. @erikd - can you try using something like https://www.stackage.org/nightly-2016-06-10/cabal.config which pins down absolutely all version ranges and see what happens? The Travis test suite seemingly passed with stack nightly.

**Comment by ndmitchell** [comment]
I just did a completely fresh Windows VM, GHC 8.0.1, update, test and it worked perfectly.

**Comment by erikd** [comment]
I grabbed that `cabal.config` file, removed the `criterion` entry so I could build criterion's git HEAD and that failed too, just like before.

**Comment by ndmitchell** [comment]
I just ran using stack nightly, on Travis, so a clean install plus Linux and the same versions as you, and it worked: https://travis-ci.org/ndmitchell/criterion/jobs/136701986.

I guess someone who really uses Linux for real should probably give it a go - my simulations might be just a little too weird.

@erikd Could you upload the sanity.json that results from your test? Mine is at https://gist.github.com/ndmitchell/2c5e4403361943e9812243a6fec3df24

**Comment by erikd** [comment]
Mine is at https://gist.github.com/erikd/a8039725d22e96a484e41d88e3370d90

One thing iI notice is that yours is a s single line while mine is on three lines.

**Comment by ndmitchell** [comment]
Yours seems to have inserted a lot more whitespace, e.g. it starts `[ "criterio"` while mine starts `["criterio"`. Furthermore, while mine is valid JSON, yours, when pasted into Chrome, gives `Uncaught SyntaxError: Unexpected token ]`.

I increasingly suspect this is nothing to do with my patch, merely version number differences that happened to show up simultaneously. Can you try a commit before my changes went in?

**Comment by erikd** [comment]
You are right. I checked out commit `0f59bcb7` (from 2014??) and that is indeed fine. I suspect this may be an issue with the JS libraries shipped with Debian.

**Comment by erikd** [comment]
The Debian versions of the relevant JS libraries are as follows:

```
$ dpkg -l | grep jquery
ii  libjs-jquery      1.12.3-1       all  JavaScript library for dynamic web applications
ii  libjs-jquery-flot 0.8.3+dfsg-1   all  plotting library for jQuery
ii  libjs-jquery-ui   1.10.1+dfsg-1  all  JavaScript UI library for dynamic web applications
```

**Comment by erikd** [comment]
I also tried reverting commits 80f78aaa97acc9e123 and 8995296ed52527ce4c7 and that also solves the problem.

I suspect the version of libjs-jquery (1.12.3) is too old. The version of the file you  removed in commit 80f78aaa97acc9e123fa8 was 2.1.1.

**Comment by ndmitchell** [comment]
jquery 1.12.3 is actually newer than 2.1.1 - it's equivalent to 2.2.3. The 1.10 vs 2.0 difference indicates whether the jquery maintains IE6-8 compatibility, but otherwise they are equivalent. There are lots of weird things here:

* Doesn't happen on my Linux or Windows tests.
* The Haskell test, which writes a JSON output file, is impacted by jQuery stuff. I can't see why that code is even invoked. I can't even see _how_ it could be invoked, without a Javascript interpreter.
* You end up with invalid JSON. That's just weird. Tracing the code path it goes nowhere near the jquery files at all - it just slaps out as JSON encoded value.

**Comment by ndmitchell** [comment]
OK, I can now reproduce it. Firstly, it's non-deterministic - I can currently get it happening about every other run. Secondly, it has nothing to do with js-jquery - completely removing those libraries still results in corruption. Current culprit is templating library or text builder.

**Comment by erikd** [comment]
> OK, I can now reproduce it. 

Yay!

> Firstly, it's non-deterministic

I was suspecting that, but hadn't managed to confirm it yet.

**Comment by ndmitchell** [comment]
OK, I think I've tracked it down... Should have it confirmed in 15 minutes or so.

**Comment by ndmitchell** [comment]
Patch in #108. js-jquery was completely irrelevant, as were package versions. What _was_ relevant was running the test multiple times, first producing a "longish" file then a "shortish" file. Pretty much guaranteed if you ran the test 5 times.

I had visions of buffer overflows, but nothing to do with that at all.

**Comment by erikd** [comment]
:+1: @ndmitchell !!!

**Comment by erikd** [comment]
Closing this. Fixed in #107 .
